File: experiments/evm_feasibility/mcopy_reference.py
#!/usr/bin/env python3
"""Reproduce the dual-pin Taraxa Ficus/Cacti MCOPY corpus."""

import argparse
import hashlib
import json
import os
import pathlib
import subprocess
import tempfile

from reference import REVISIONS, ROOT

HERE = pathlib.Path(__file__).resolve().parent
FIXTURES = HERE / "fixtures" / "mcopy"


def fail(message):
    raise RuntimeError(message)


def checked_output(command, **kwargs):
    try:
        return subprocess.check_output(command, **kwargs)
    except subprocess.CalledProcessError as error:
        fail(f"command failed ({error.returncode}): {' '.join(map(str, command))}")


def validate_artifact(label, data):
    try:
        document = json.loads(data)
    except (UnicodeDecodeError, json.JSONDecodeError) as error:
        fail(f"{label} emitted invalid JSON: {error}")
    rows = document.get("mcopy")
    if not isinstance(rows, list) or not rows:
        fail(f"{label} omitted MCOPY rows")
    names = set()
    for index, row in enumerate(rows):
        if not isinstance(row, dict):
            fail(f"{label} row {index} is not an object")
        required = {
            "name",
            "phase",
            "code",
            "gas_used",
            "output",
            "execution_error",
            "outer_error",
        }
        if set(row) != required:
            fail(f"{label} row {index} has unexpected fields")
        if row["name"] in names:
            fail(f"{label} duplicate row name: {row['name']}")
        names.add(row["name"])
        if row["phase"] not in ("californicum", "ficus", "cacti"):
            fail(f"{label} row {index} has invalid phase")
        for field in ("code", "output"):
            try:
                bytes.fromhex(row[field])
            except (TypeError, ValueError) as error:
                fail(f"{label} row {index} invalid {field}: {error}")
        if not isinstance(row["gas_used"], int) or not 0 <= row["gas_used"] <= 100000:
            fail(f"{label} row {index} has invalid gas")
        if not isinstance(row["execution_error"], str) or not isinstance(
            row["outer_error"], str
        ):
            fail(f"{label} row {index} has invalid error")
    needed = {
        "before-ficus",
        "ficus-zero-length-wide-dst",
        "cacti-inherits-zero-length-wide-dst",
        "ficus-copy-forward-overlap",
        "ficus-copy-backward-overlap",
        "cacti-copy-forward-overlap",
        "ficus-expand-empty-memory",
        "ficus-copy-two-words",
        "ficus-expansion-out-of-gas",
        "ficus-stack-underflow",
    }
    missing = needed - names
    if missing:
        fail(f"{label} omitted required cases: {sorted(missing)}")


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--record", action="store_true")
    args = parser.parse_args()
    source = pathlib.Path(os.environ.get("TARAXA_EVM_SOURCE", ROOT / "submodules/taraxa-evm"))
    exporter = (HERE / "mcopy_reference.go").read_bytes()
    artifacts = {}
    for label, revision in REVISIONS.items():
        with tempfile.TemporaryDirectory(prefix="rustaxa-mcopy-") as directory:
            tree = pathlib.Path(directory)
            archive = checked_output(["git", "-C", str(source), "archive", revision])
            subprocess.run(["tar", "-x", "-C", directory], input=archive, check=True)
            command = tree / "cmd/mcopy_reference"
            command.mkdir(parents=True)
            (command / "mcopy_reference.go").write_bytes(exporter)
            data = checked_output(
                ["go", "run", "-mod=readonly", "./cmd/mcopy_reference"], cwd=tree
            )
            validate_artifact(label, data)
            artifacts[label] = data
    if artifacts["public"] != artifacts["local"]:
        fail("pinned Go MCOPY outputs disagree")

    manifest = {
        "schema": 1,
        "references": REVISIONS,
        "go_version": checked_output(["go", "version"], text=True).strip(),
        "scope": (
            "EVM-level Californicum/Ficus/Cacti MCOPY activation, overlap, memory "
            "expansion, zero-length wide offset, stack and gas boundaries; no state, "
            "native registry, publication or production-routing claim"
        ),
        "transaction_gas_limit": 100000,
        "transaction_intrinsic_gas": 21000,
        "sha256": {
            label: hashlib.sha256(data).hexdigest() for label, data in artifacts.items()
        },
        "exporter_sha256": hashlib.sha256(exporter).hexdigest(),
    }
    expected = {
        "public.json": artifacts["public"],
        "local.json": artifacts["local"],
        "manifest.json": (json.dumps(manifest, indent=2) + "\n").encode(),
    }
    if args.record:
        FIXTURES.mkdir(parents=True, exist_ok=True)
        for name, data in expected.items():
            (FIXTURES / name).write_bytes(data)
        return
    for name, data in expected.items():
        path = FIXTURES / name
        if not path.is_file():
            fail(f"missing fixture: {path}; run with --record after review")
        if path.read_bytes() != data:
            fail(f"fixture differs: {path}; run with --record only for an intentional update")


if __name__ == "__main__":
    main()

File: experiments/evm_feasibility/mcopy_reference.go
// Pinned EVM-level oracle for Taraxa's Ficus/Cacti MCOPY instruction.
package main

import (
	"encoding/hex"
	"encoding/json"
	"math/big"
	"os"

	"github.com/Taraxa-project/taraxa-evm/common"
	"github.com/Taraxa-project/taraxa-evm/core/vm"
	"github.com/Taraxa-project/taraxa-evm/crypto"
	"github.com/Taraxa-project/taraxa-evm/params"
	"github.com/Taraxa-project/taraxa-evm/taraxa/state/state_db"
	"github.com/Taraxa-project/taraxa-evm/taraxa/state/state_evm"
)

type mcopyInput struct {
	sender common.Address
	target common.Address
	code   []byte
}

func (i mcopyInput) GetCode(hash *common.Hash) []byte {
	if crypto.Keccak256Hash(i.code) != *hash {
		panic("unexpected code hash")
	}
	return clone(i.code)
}

func (i mcopyInput) GetAccount(address *common.Address, callback func(state_db.Account)) {
	if *address == i.sender {
		callback(state_db.Account{Nonce: big.NewInt(1), Balance: big.NewInt(1_000_000)})
		return
	}
	if *address == i.target {
		hash := crypto.Keccak256Hash(i.code)
		callback(state_db.Account{
			Nonce: big.NewInt(1), Balance: big.NewInt(0), CodeHash: &hash, CodeSize: uint64(len(i.code)),
		})
	}
}

func (mcopyInput) GetAccountStorage(*common.Address, *common.Hash, func([]byte)) {
	panic("unexpected storage read")
}

type mcopyRow struct {
	Name           string `json:"name"`
	Phase          string `json:"phase"`
	Code           string `json:"code"`
	GasUsed        uint64 `json:"gas_used"`
	Output         string `json:"output"`
	ExecutionError string `json:"execution_error"`
	OuterError     string `json:"outer_error"`
}

func clone(value []byte) []byte { return append([]byte(nil), value...) }

func push(code []byte, value *big.Int) []byte {
	bytes := value.Bytes()
	if len(bytes) == 0 {
		bytes = []byte{0}
	}
	if len(bytes) > 32 {
		panic("EVM push exceeds one word")
	}
	return append(append(code, byte(0x5f+len(bytes))), bytes...)
}

func mcopyProgram(initial []byte, dst, src, length *big.Int, returnSize uint64) []byte {
	code := []byte{}
	if initial != nil {
		if len(initial) != 32 {
			panic("initial MCOPY word must contain 32 bytes")
		}
		code = push(code, new(big.Int).SetBytes(initial))
		code = push(code, new(big.Int))
		code = append(code, 0x52) // MSTORE
	}
	code = push(code, length)
	code = push(code, src)
	code = push(code, dst)
	code = append(code, 0x5e) // MCOPY
	code = push(code, new(big.Int).SetUint64(returnSize))
	code = push(code, new(big.Int))
	return append(code, 0xf3) // RETURN
}

func execute(name, phase string, code []byte) mcopyRow {
	sender := common.BytesToAddress([]byte{0xaa})
	target := common.BytesToAddress([]byte{0xbb})
	state := new(state_evm.TransitionState)
	state.Init(state_evm.Opts{})
	state.SetInput(mcopyInput{sender: sender, target: target, code: code})
	var evm vm.EVM
	evm.Init(
		func(uint64) *big.Int { return new(big.Int) },
		state,
		vm.DefaultOpts(),
		params.TestChainConfig,
		vm.Config{},
	)
	rules := vm.Rules{IsCornus: true}
	switch phase {
	case "californicum":
	case "ficus":
		rules.IsFicus = true
	case "cacti":
		rules.IsFicus = true
		rules.IsCacti = true
	default:
		panic("unknown MCOPY phase")
	}
	evm.SetBlock(
		&vm.Block{Number: 1, BlockInfo: vm.BlockInfo{GasLimit: 1_000_000, Difficulty: new(big.Int)}},
		rules,
	)
	result, err := evm.Main(&vm.Transaction{
		From: sender, To: &target, Nonce: big.NewInt(1), GasPrice: big.NewInt(1),
		Value: new(big.Int), Gas: 100_000,
	})
	outerError := ""
	if err != nil {
		outerError = err.Error()
	}
	return mcopyRow{
		Name: name, Phase: phase, Code: hex.EncodeToString(code), GasUsed: result.GasUsed,
		Output: hex.EncodeToString(result.CodeRetval), ExecutionError: string(result.ExecutionErr),
		OuterError: outerError,
	}
}

func main() {
	pattern := make([]byte, 32)
	for i := range pattern {
		pattern[i] = byte(i)
	}
	zero := new(big.Int)
	one := big.NewInt(1)
	sixteen := big.NewInt(16)
	thirtyTwo := big.NewInt(32)
	thirtyThree := big.NewInt(33)
	wide := new(big.Int).Lsh(big.NewInt(1), 255)
	large := new(big.Int).SetUint64(1 << 20)

	cases := []struct {
		name  string
		phase string
		code  []byte
	}{
		{"before-ficus", "californicum", mcopyProgram(nil, zero, zero, zero, 0)},
		{"ficus-zero-length-wide-dst", "ficus", mcopyProgram(nil, wide, zero, zero, 0)},
		{"cacti-inherits-zero-length-wide-dst", "cacti", mcopyProgram(nil, wide, zero, zero, 0)},
		{"ficus-copy-forward-overlap", "ficus", mcopyProgram(pattern, one, zero, sixteen, 32)},
		{"ficus-copy-backward-overlap", "ficus", mcopyProgram(pattern, zero, one, sixteen, 32)},
		{"cacti-copy-forward-overlap", "cacti", mcopyProgram(pattern, one, zero, sixteen, 32)},
		{"ficus-expand-empty-memory", "ficus", mcopyProgram(nil, sixteen, thirtyTwo, one, 64)},
		{"ficus-copy-two-words", "ficus", mcopyProgram(pattern, thirtyTwo, zero, thirtyThree, 96)},
		{"ficus-expansion-out-of-gas", "ficus", mcopyProgram(nil, large, zero, one, 0)},
		{"ficus-stack-underflow", "ficus", []byte{0x5e, 0x00}},
	}
	rows := make([]mcopyRow, 0, len(cases))
	for _, item := range cases {
		rows = append(rows, execute(item.name, item.phase, item.code))
	}
	if err := json.NewEncoder(os.Stdout).Encode(map[string]any{"mcopy": rows}); err != nil {
		panic(err)
	}
}
